### ISSUE TITLE
Performance claims

### ISSUE THREAD

**Issue by spendergrsec** [IAIK/KAISER#3]
Hi,

I was wondering if during peer review of this paper, it was ever asked why no micro-benchmarks or other more representative benchmarks were performed (or if they were performed, why they were not presented)?  I've read many academic security papers that make performance claims, and I'm not sure I can recall ever seeing the set you've chosen to form the basis of your claims.

For instance, why did you choose not to run lmbench?

My own tests with an apples-apples comparison with your patch on an Intel i7-5600U (same kernel config, same system, same GCC, etc -- only difference being the patch) shows that on a du -s test (the same we performed for RAP) KAISER has a 70% hit on kernel time, and a 100% hit on overall execution time (since KAISER affects both userland and kernel execution time).

It's clear what the reasons for this discrepancy are, but I'm curious if you can explain your performance testing methodology, since it'd be easy to hide actual performance impact by specifically choosing benchmarks that spent the majority of their time in userland computation with very little kernel re-entry.

Thanks,
-Brad

**Comment by dgruss** [comment]
There are many benchmarks, we chose these three because we think they are representative, we did not perform others because our results were quite consistent over multiple benchmarks and also on some smaller experiments. PARSEC is an open source benchmark suite which is commonly used http://parsec.cs.princeton.edu/. It should also do quite a lot of syscalls...

I believe that there must be some performance difference on different architectures or even processor models. Clearly, we take advantage of the size of the TLB and also the "when" TLB entries are invalidated. This may be different on different models. For instance, I would assume that a Nehalem processor would take a significant performance hit because it has much fewer optimizations around the TLB.

**Comment by spendergrsec** [comment]
I've done another set of performance tests, stricter than the previous ones (single user mode, locked to a single core, HT disabled, speedstep disabled) on the same i7 6700 CPU I used for the RAP benchmarks, performing the same du -s test as previously (with perf stat -r 10 du -s).

The relevant raw perf results are:
KASIER-on:
4022.214001 task-clock (msec)
13,641,008,493 cycles
10,997,430,939 instructions
2,070,222,999 branches
29,769,985 branch-misses
0.80 insn per cycle

KAISER-off:
2486.725311 task-clock (msec)
8,431,879,087 cycles
10,850,975,460 instructions
2,071,413,447 branches
29,766,611 branch-misses
1.29 insn per cycle

So a 62% hit for the same test on the same CPU architecture where you reported a < 1% hit.

I have tried to perform LMBENCH benchmarks for KAISER, but cannot successfully complete the test despite numerous (5+) attempts -- partway through the benchmarking process (during the networking tests) the system experiences a hard reset.  This happens only on the KAISER-enabled kernel.

Several of the tests were able to run fully before the hard reset, and are reproduced below:
KAISER-on (in microseconds, hit over KASIER-off in parens):
simple syscall: 0.9077 (541%)
simple read: 0.3773  (271%)
simple write: 0.3107 (365%)
simple stat: 0.5827 (128%)
simple fstat: 0.3520 (269%)
simple open/close: 1.3378 (88%)
select on 10 fd's: 0.4775 (124%)
select on 100 fd's: 1.0114 (49%)
select on 250 fd's: 1.8673 (36%)
select on 500 fd's: 3.3308 (21%)
select on 10 tcp fd's: 0.5209 (119%)
select on 100 tcp fd's: 2.1326 (9%)
select on 250 tcp fd's: 4.7934 (0%)
select on 500 tcp fd's: 9.8232 (10%)
signal handler installation: 0.3577 (195%)
signal handler overhead: 1.229 (48%)
protection fault: 0.6185 (81%)
pipe latency: 4.4233 (17%)
AF_UNIX sock stream latency: 5.0346 (51%)
Process fork+exit: 89.8197 (15%)
Process fork+execve: 91.7500 (10%)

KASIER-off (in microseconds):
simple syscall: 0.1416
simple read: 0.1017
simple write: 0.0668
simple stat: 0.2554
simple fstat: 0.0954
simple open/close: 0.7126
select on 10 fd's: 0.2134
select on 100 fd's: 0.6775
select on 250 fd's: 1.3779
select on 500 fd's: 2.7578
select on 10 tcp fd's: 0.2376
select on 100 tcp fd's: 1.9487
select on 250 tcp fd's: 4.8119
select on 500 tcp fd's: 8.9303
Signal handler installation: 0.1213
Signal handler overhead: 0.8332
Protection fault: 0.3413
Pipe latency: 3.7739
AF_UNIX sock stream latency: 3.3451
Process fork+exit: 78.4030
Process fork+execve: 83.1515

Sorry, but KASLR is still a DOA mitigation, and cooking the benchmarking books won't revive it.  Based on the results above, people should be very suspicious of performance claims made in your future papers: academic research should be science, not politics.

-Brad

**Comment by dgruss** [comment]
thank you for your comment. we did not run your du -s test and we also did not claim. we repeated several of our experiments, to make sure the numbers we reported are indeed reproducible, getting the same numbers as in our paper again. the benchmarks we used are commonly used both in academic research and also to compare system performance as they are expected to represent realistic workloads. we will take a look at more benchmarks like the ones you suggested...

**Comment by dgruss** [comment]
Hi Brad,

I just did some tests on my private machine here and get similar results (see below). However, I'm not sure what the conclusion is here... simple syscall seems to be a bit off, in our micro benchmarks we observed an overhead for a simple syscall of less than 300 cycles (to switch cr3).
also this is clearly not what the other benchmarks aiming at realistic workloads present.
I'm also not sure what to think of the cases where the kernel with kaiser was faster (fork+exit is supposed to be much slower while fork+execve is faster!? i don't really believe this... - also note a similar thing in your numbers for the 250 fds case - not sure where this comes from).
What becomes visible is that for larger numbers we indeed have low overheads.
However, this was just a single run (no crash) with and without kaiser (also note the different kernel versions because i didn't have the exact same one at hand right now).

                    4.10.0-rc6-kaiser    4.10.0-20-generic    overhead
Simple syscall            0,8729            0,0342            2452,3%
Simple read                0,3002            0,0867            246,3%
Simple write                0,2641            0,0579            356,1%
Simple stat                0,5485            0,3096            77,2%
Simple fstat                0,2885            0,089            224,2%
Simple open/close            1,1025            0,7432            48,3%
Select on 10 fd's            0,4084            0,1964            107,9%
Select on 100 fd's            0,8475            0,7104            19,3%
Select on 250 fd's            1,4767            1,3604            8,5%
Select on 500 fd's            2,6634            2,6243            1,5%
Select on 10 tcp fd's            0,4559            0,2218            105,5%
Select on 100 tcp fd's        1,755            1,5323            14,5%
Select on 250 tcp fd's        3,918            4,3234            -9,4%
Select on 500 tcp fd's        7,6887            7,3849            4,1%
Signal handler installation        0,3085            0,1088            183,5%
Signal handler overhead        0,8931            0,6678            33,7%
Protection fault            0,5445            0,283            92,4%
Pipe latency                3,9756            3,3529            18,6%
AF_UNIX sock stream latency        4,583            3,3265            37,8%
Process fork+exit            58,4066            50,2857            16,1%
Process fork+execve            63,0244            63,7273            -1,1% 


Anyway, the performance claim from the paper is a low overhead based on another set of benchmarks (which aims to create realistic workloads) and for that set of benchmark we still get the same performance results. Naturally we will mention all the insights we gained while we made our patch public including the overheads you measured and that we got reported from several people from the LKML. This will allow for a more complete picture of what is possible and what it costs.


Cheers,
Daniel

**Comment by spendergrsec** [comment]
Did the vanilla kernel in your tests use some Ubuntu-patched version?  The kernel version isn't what I'd expect for an apples->apples comparison (even if likely similar).  Was the same exact kernel config used between both versions?  As for the oddities, if the tests weren't done strictly as I ran them, then you end up with more variability (speedstep, turbo, HT, other activity, etc).  Small percent differences like that on wall time are generally noise (run multiple times / generate stddev to confirm) and usually require perf running the tests multiple times to discern any real performance hit.

-Brad

**Comment by dgruss** [comment]
As I said, that was just what I had at hand here now and numbers should 
be taken with care... Should be same kernel config, but it's obviously a 
different kernel version ;)

Disabling cpu features is also not really the right way when you want to 
measure real world performance.
For the paper we performed a thorough evaluation with a significant 
number of runs of the benchmark suites. I agree with you, multiple runs 
is the right way to measure the real performance hit.


Cheers,
Daniel

**Comment by misc0110** [comment]
For the paper we had the same kernel version with and without KAISER enabled, all other options were the same. As kernel config, we used the Ubuntu config to run Ubuntu 16.10.

Running a full Ubuntu clearly introduces some noise, but we chose this to represent a real world setting. 
To reduce the noise, we ran the benchmarks multiple times. I don't have all the number to hand right now, but e.g. for the single threaded PARSEC 3.0 benchmark, we had 10 runs for both the kernel with and without KAISER. Each of the runs took approximately 28 minutes wall time, resulting in a total benchmarking time of around 4 hours and 40 minutes for each kernel version. We had an average difference of 4 seconds per run (=0.37%). The standard deviation of the runs is 2.54 seconds, which should be low enough to ensure that the numbers are not due to noise.

Your benchmarks are quite fast (in the range of nanoseconds to microseconds), thus small variations in the runtime already result in a high variance of the percentage (in both directions). 
We wanted KAISER to be a practical approach that can be enabled in hardened Linux kernels. Thus, we were especially interested in real world performance, and not in artificial stress tests. I would expect that it is possible to find scenarios where the KAISER-enabled kernel performs a lot worse than the vanilla kernel.  

I'm still thankful for your efforts in trying to reproduce the performance results. lmbench is now also on our list for benchmarks in future works.

Michael
